File: packages/desktop/test/vault-app.test.ts
import { describe, it, expect } from "vitest";
import { VaultApp } from "../src/core/vault-app.js";
import { MemoryStore, FileStore } from "../src/core/local-store.js";
import type { Transport, RemoteItem, PushOutcome } from "../src/core/sync-client.js";
import { generateSalt } from "@vaultkeep/crypto";
import { VaultService } from "@vaultkeep/server/vault/vault.service.js";
import { InMemoryVaultRepository } from "@vaultkeep/server/vault/vault.repository.js";

const FAST_KDF = { memoryKiB: 8192, iterations: 2, parallelism: 1 };

/**
 * Bridge the real server VaultService to the desktop Transport interface, so
 * the desktop app talks to the ACTUAL sync engine (real conflict resolution),
 * just without an HTTP socket in the way.
 */
function serverTransport(svc: VaultService, userId: string): Transport {
  return {
    async pull(since?: string): Promise<RemoteItem[]> {
      return svc.pull(userId, since);
    },
    async push(id, ciphertext, baseVersion): Promise<PushOutcome> {
      const r = await svc.push(userId, { id, ciphertext, baseVersion });
      if (r.status === "ok") return { status: "ok", version: r.version };
      return { status: "conflict", serverVersion: r.serverVersion, server: r.server };
    },
  };
}

describe("VaultApp end-to-end", () => {
  it("rejects unlock with the wrong master password", async () => {
    const salt = generateSalt();
    // Seed a vault encrypted under the RIGHT password.
    const store = new MemoryStore();
    const right = new VaultApp(salt, store, null, FAST_KDF);
    await right.unlock("correct-password");
    await right.add({ type: "login", title: "Amazon", password: "s3cret" });
    right.lock();

    // A different password must fail (GCM auth tag won't verify).
    const wrong = new VaultApp(salt, store, null, FAST_KDF);
    await expect(wrong.unlock("WRONG-password")).rejects.toThrow();
  });

  it("persists items encrypted at rest and reloads them", async () => {
    const salt = generateSalt();
    const store = new MemoryStore();

    const a = new VaultApp(salt, store, null, FAST_KDF);
    await a.unlock("master");
    await a.add({ type: "login", title: "GitHub", username: "henry", password: "pw1" });
    a.lock();

    // The on-disk blob must contain neither the password NOR the title/username.
    const raw = (await store.readRaw()) ?? "";
    expect(raw).not.toContain("pw1");
    expect(raw).not.toContain("GitHub");
    expect(raw).not.toContain("henry");

    // Reopen and decrypt.
    const b = new VaultApp(salt, store, null, FAST_KDF);
    await b.unlock("master");
    expect(b.list()[0].password).toBe("pw1");
  });

  it("syncs a secret from one device to another, zero-knowledge", async () => {
    const salt = generateSalt();
    const svc = new VaultService(new InMemoryVaultRepository());
    const transport = serverTransport(svc, "user-1");

    // Device A creates an item and pushes it.
    const mac = new VaultApp(salt, new MemoryStore(), transport, FAST_KDF);
    await mac.unlock("master-pw");
    await mac.add({ type: "login", title: "Bank", username: "h@x.com", password: "vault-pw" });
    const up = await mac.sync();
    expect(up.pushed).toBe(1);

    // The SERVER must only hold ciphertext — never the plaintext password.
    const stored = JSON.stringify(await svc.pull("user-1"));
    expect(stored).not.toContain("vault-pw");
    expect(stored).not.toContain("Bank");

    // Device B (same account/password, fresh local store) pulls and decrypts.
    const win = new VaultApp(salt, new MemoryStore(), transport, FAST_KDF);
    await win.unlock("master-pw");
    const down = await win.sync();
    expect(down.pulled).toBe(1);
    const item = win.list()[0];
    expect(item.title).toBe("Bank");
    expect(item.password).toBe("vault-pw");
  });

  it("resolves a cross-device conflict by taking the server's newer copy", async () => {
    const salt = generateSalt();
    const svc = new VaultService(new InMemoryVaultRepository());
    const t = serverTransport(svc, "user-2");

    const mac = new VaultApp(salt, new MemoryStore(), t, FAST_KDF);
    await mac.unlock("pw");
    const item = await mac.add({ type: "login", title: "Site", password: "v1" });
    await mac.sync(); // now at version 1 on server

    const win = new VaultApp(salt, new MemoryStore(), t, FAST_KDF);
    await win.unlock("pw");
    await win.sync(); // win learns version 1

    // Windows edits and pushes -> server v2.
    await win.update(item.id, { password: "v2-from-win" });
    expect((await win.sync()).pushed).toBe(1);

    // Mac edits the SAME item while still at v1 -> conflict, takes server copy.
    await mac.update(item.id, { password: "v2-from-mac-LOSES" });
    const res = await mac.sync();
    expect(res.conflicts).toBe(1);
    expect(mac.get(item.id)!.password).toBe("v2-from-win");
  });
});

describe("FileStore persistence (key-free, single blob)", () => {
  it("round-trips a vault through a fake filesystem", async () => {
    const salt = generateSalt();
    // A tiny in-memory fake of the fs API FileStore expects.
    const files = new Map<string, string>();
    const fs = {
      readFile: async (p: string) => {
        if (!files.has(p)) throw new Error("ENOENT");
        return files.get(p)!;
      },
      writeFile: async (p: string, d: string) => void files.set(p, d),
    };

    const store1 = new FileStore("/tmp/vault.enc", fs);
    const a = new VaultApp(salt, store1, null, FAST_KDF);
    await a.unlock("master");
    await a.add({ type: "login", title: "Proton", password: "fs-secret" });
    a.lock();

    // The bytes written to "disk" leak nothing.
    expect(files.get("/tmp/vault.enc")).not.toContain("fs-secret");
    expect(files.get("/tmp/vault.enc")).not.toContain("Proton");

    // A brand-new FileStore over the same fake fs reloads and decrypts.
    const b = new VaultApp(salt, new FileStore("/tmp/vault.enc", fs), null, FAST_KDF);
    await b.unlock("master");
    expect(b.list()[0].password).toBe("fs-secret");
  });
});

// BiometricUnlock has its own suite in test/biometric.test.ts (wrapped-key
// design: enroll/unlock, declined prompt, unavailable hardware, tampered or
// stale wrapped keys, unenroll wiping the keystore).

File: packages/mobile/test/vault_app_test.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:cryptography/cryptography.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:vaultkeep_mobile/core/local_store.dart';
import 'package:vaultkeep_mobile/core/sync_client.dart';
import 'package:vaultkeep_mobile/core/vault_app.dart';
import 'package:vaultkeep_mobile/core/vault_crypto.dart';

// Fast KDF for tests — algorithm identical, cost turned down.
const fastKdf = KdfParams(memoryKiB: 256, iterations: 1, parallelism: 1);
const salt = 'q83vASNFZ4mrze8BI0VniQ==';

/// In-memory server double implementing the real optimistic-concurrency
/// contract (baseVersion must match, conflicts return the server copy).
class FakeServer implements Transport {
  final Map<String, RemoteItem> rows = {};
  @override
  Future<List<RemoteItem>> pull({String? since}) async => rows.values.toList();
  @override
  Future<PushOutcome> push(String id, String ciphertext, int? baseVersion) async {
    final cur = rows[id];
    final curVersion = cur?.version ?? 0;
    if ((baseVersion ?? 0) != curVersion) {
      return PushOutcome(status: 'conflict', server: cur);
    }
    final next = RemoteItem(
        id: id,
        ciphertext: ciphertext,
        version: curVersion + 1,
        updatedAt: DateTime.now().toIso8601String());
    rows[id] = next;
    return PushOutcome(status: 'ok', version: next.version);
  }
}

void main() {
  test('unlock → add → lock → reopen with the same password', () async {
    final store = MemoryStore();
    final a = VaultApp(salt, store, null, kdf: fastKdf);
    await a.unlock('master pw 1');
    await a.add(title: 'GitHub', username: 'henry', password: 'gh-secret');
    a.lock();
    expect(() => a.list(), throwsStateError);

    final b = VaultApp(salt, store, null, kdf: fastKdf);
    await b.unlock('master pw 1');
    expect(b.list().single.password, 'gh-secret');
  });

  test('a wrong password fails via the GCM auth tag', () async {
    final store = MemoryStore();
    final a = VaultApp(salt, store, null, kdf: fastKdf);
    await a.unlock('right password');
    await a.add(title: 'X', password: 'p');
    a.lock();

    final b = VaultApp(salt, store, null, kdf: fastKdf);
    expect(() => b.unlock('wrong password'),
        throwsA(isA<SecretBoxAuthenticationError>()));
    expect(b.isUnlocked, false);
  });

  test('the on-disk blob is one opaque ciphertext — no plaintext leaks',
      () async {
    final store = MemoryStore();
    final a = VaultApp(salt, store, null, kdf: fastKdf);
    await a.unlock('master pw 1');
    await a.add(title: 'Proton', username: 'henry', password: 'super-secret');
    final raw = (await store.readRaw())!;
    expect(raw, isNot(contains('Proton')));
    expect(raw, isNot(contains('super-secret')));
    expect(() => jsonDecode(raw), throwsFormatException); // base64, not JSON
  });

  test('two-way sync: push, pull on a second device, conflict adopts server',
      () async {
    final server = FakeServer();

    final phone = VaultApp(salt, MemoryStore(), server, kdf: fastKdf);
    await phone.unlock('shared pw');
    final item = await phone.add(title: 'Bank', password: 'v1');
    expect((await phone.sync()).pushed, 1);

    // A second device pulls the item and edits it FIRST (wins the race).
    final laptop = VaultApp(salt, MemoryStore(), server, kdf: fastKdf);
    await laptop.unlock('shared pw');
    expect((await laptop.sync()).pulled, 1);
    await laptop.update(item.id, password: 'v2-from-laptop');
    expect((await laptop.sync()).pushed, 1); // server now at version 2

    // The phone edits the same item from its stale version → conflict →
    // the server's newer copy is adopted (same policy as desktop).
    await phone.update(item.id, password: 'v2-from-phone');
    await phone.add(title: 'Unrelated', password: 'x');
    final s = await phone.sync();
    expect(s.conflicts, 1);
    expect(s.pushed, 1); // the unrelated new item still went up
    expect(phone.list().firstWhere((i) => i.id == item.id).password,
        'v2-from-laptop');

    // Everyone converges after one more pull.
    await laptop.sync();
    expect(laptop.list().length, 2);
  });

  test('snapshotKey returns an independent copy of the live master key',
      () async {
    final app = VaultApp(salt, MemoryStore(), null, kdf: fastKdf);
    final key = await deriveMasterKey('pw', salt, fastKdf);
    await app.unlockWithKey(key);
    final snapshot = app.snapshotKey();
    expect(snapshot, key);
    snapshot.fillRange(0, snapshot.length, 0);
    // Mutating the snapshot must not zero the vault's own live key.
    expect(key.every((b) => b == 0), false);
  });

  group('rekey (recovery-key password reset)', () {
    test(
        're-encrypts the local blob and re-pushes every item under the new key',
        () async {
      final server = FakeServer();
      final store = MemoryStore();
      final app = VaultApp(salt, store, server, kdf: fastKdf);
      await app.unlock('old password');
      final item =
          await app.add(title: 'GitHub', username: 'henry', password: 'gh-secret');
      await app.sync();
      final versionBefore = server.rows[item.id]!.version;

      final newKey = await deriveMasterKey('new password', salt, fastKdf);
      await app.rekey(newKey);

      // The item was marked dirty by rekey, so syncing again re-pushes it —
      // the server's old copy is ciphertext under a key nobody has anymore.
      final s = await app.sync();
      expect(s.pushed, 1);
      expect(server.rows[item.id]!.version, greaterThan(versionBefore));

      // A fresh app unlocked with the NEW key reads the same item back.
      final reopened = VaultApp(salt, store, server, kdf: fastKdf);
      await reopened.unlockWithKey(newKey);
      expect(reopened.list().single.password, 'gh-secret');

      // The OLD password can no longer open the (now re-keyed) local blob.
      final stale = VaultApp(salt, store, server, kdf: fastKdf);
      expect(() => stale.unlock('old password'),
          throwsA(isA<SecretBoxAuthenticationError>()));
    });

    test('rekeying a locked vault throws', () async {
      final app = VaultApp(salt, MemoryStore(), null, kdf: fastKdf);
      final key = await deriveMasterKey('x', salt, fastKdf);
      expect(() => app.rekey(key), throwsStateError);
    });

    test('rejects a malformed replacement key', () async {
      final app = VaultApp(salt, MemoryStore(), null, kdf: fastKdf);
      await app.unlock('pw');
      expect(() => app.rekey(Uint8List(16)), throwsArgumentError);
    });
  });
}
